pkg/codegen/go/gen_program.go

```diff
@@ -159,21 +182,41 @@ func (g *generator) genResource(w io.Writer, r *hcl2.Resource) {
 		g.genTemps(w, temps)
 	}
 
-	g.Fgenf(w, "%s, err := %s.New%s(ctx, \"%[1]s\", ", resName, mod, typ)
-	if len(r.Inputs) > 0 {
-		g.Fgenf(w, "&%s.%sArgs{\n", mod, typ)
-		for _, attr := range r.Inputs {
-			g.Fgenf(w, "%s: ", strings.Title(attr.Name))
-			g.Fgenf(w, "%.v,\n", attr.Value)
+	instantiate := func(varName, resourceName string) {
+		if g.scopeTraversalRoots.Has(varName) {
+			g.Fgenf(w, "%s, err := %s.New%s(ctx, %s, ", varName, mod, typ, resourceName)
+		} else {
+			g.Fgenf(w, "_, err = %s.New%s(ctx, %s, ", mod, typ, resourceName)
+		}
 
+		if len(r.Inputs) > 0 {
+			g.Fgenf(w, "&%s.%sArgs{\n", mod, typ)
+			for _, attr := range r.Inputs {
+				g.Fgenf(w, "%s: ", strings.Title(attr.Name))
+				g.Fgenf(w, "%.v,\n", attr.Value)
+
+			}
+			g.Fgenf(w, "})\n")
+		} else {
+			g.Fgenf(w, "nil)\n")
 		}
-		g.Fgenf(w, "})\n")
+		g.Fgenf(w, "if err != nil {\n")
+		g.Fgenf(w, "return err\n")
+		g.Fgenf(w, "}\n")
+	}
+
+	if r.Options != nil && r.Options.Range != nil {
+		rangeType := model.ResolveOutputs(r.Options.Range.Type())
+		rangeExpr, temps := g.lowerExpression(r.Options.Range, rangeType, false)
+		g.genTemps(w, temps)
+
+		g.Fgenf(w, "for i0, val0 := range %.v {\n", rangeExpr)
```

Comment:
Should `i0` be `key0` to align with the code in `genScopeTraversalExpression`?

Reply:
yup, In the next PR I updated the traversal code to reference i0, but you're right that for both objects/arrays key makes more sense. I'll update all usages.